apps/settings/js/modules/feature_model.js

```diff
@@ -0,0 +1,99 @@
+define(function(require) {
+  'use strict';
+
+  var Module = require('modules/base/module');
+  var Observable = require('modules/mvvm/observable');
+
+  /*
+   * A model needs to be created with an options object with the following
+   * three functions:
+   *
+   * onInit is called once with a function as the first argument when the
+   * model is initializing. The function is used for reporting the state
+   * changes. The function should return a initial state immediately.
+   *
+   * onSetState is called with a desired state. Do the operation required for
+   * switching to the desired state and return a promise wrapping the state
+   * after the operation no matter it resolves or rejects.
+   *
+   * onGetState is called for requesting the current state.
+   *
+   * @class FeatureModel
+   * @param {Object} options
+   * @param {Function} options.onInit
+   * @param {Function} options.onSetState
+   * @param {Function} options.onGetState
+   * @returns {FeatureModel}
+   */
+  var FeatureModel = Module.create(function FeatureModel(options) {
+    this.super(Observable).call(this);
+
+    this._init(options);
+  }).extend(Observable);
+
+  Observable.defineObservableProperty(FeatureModel.prototype, 'targetState', {
+    value: null
+  });
+
+  Observable.defineObservableProperty(FeatureModel.prototype, 'currentState', {
+    readonly: true,
+    value: null
+  });
+
+  Observable.defineObservableProperty(FeatureModel.prototype, 'transitioning', {
+    readonly: true,
+    value: false
+  });
+
+  FeatureModel.prototype._init = function(options) {
+    // XXX: do more checks here
+    this._options = options;
+
+    this.targetState = this._options.onInit((realState) => {
+      if (!this._transitioning) {
+        this._currentState = realState;
+        this.targetState = realState;
+      }
+    });
+
+    // Do initialization
+    this._options.onGetState().then((realState) => {
+      this._currentState = realState;
+      if (this.targetState !== realState) {
+        // If the current target state is not the real state, do set again.
+        this._setState(this.targetState);
+      }
+
+      this.observe('targetState', (newState) => {
+        if (this._transitioning || newState === this._currentState) {
+          return;
+        }
+        this._setState(newState);
+      });
+    }).catch((err) => {
+      this.error('initialization error ' + err);
+    });
+  };
+
+  FeatureModel.prototype._setState = function(state) {
+    this.debug('set state: ' + state);
+    this._transitioning = true;
+    return this._options.onSetState(state).then((realState) => {
+      this.debug('set state completed: ' + state +
+        ', real state: ' + realState);
+      this._transitioning = false;
+      this._currentState = realState;
+      if (this.targetState !== realState) {
+        // If the current target state is not the real state, do set again.
+        this._setState(this.targetState);
+      }
+    }, (realState) => {
```

Comment:
interesting, if the request failed, how can it return a `realState` here ? I thought this should be `error` ?